addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc1/NikoHomeControlCommunication1.java

```diff
@@ -0,0 +1,542 @@
+import org.openhab.binding.nikohomecontrol.internal.protocol.NhcAction;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NhcControllerEvent;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NhcThermostat;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NikoHomeControlCommunication;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NikoHomeControlConstants.ActionType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link NikoHomeControlCommunication1} class is able to do the following tasks with Niko Home Control I
+ * systems:
+ * <ul>
+ * <li>Start and stop TCP socket connection with Niko Home Control IP-interface.
+ * <li>Read all setup and status information from the Niko Home Control Controller.
+ * <li>Execute Niko Home Control commands.
+ * <li>Listen to events from Niko Home Control.
+ * </ul>
+ *
+ * @author Mark Herwege - Initial Contribution
+ */
+@NonNullByDefault
+public class NikoHomeControlCommunication1 extends NikoHomeControlCommunication {
+
+    private Logger logger = LoggerFactory.getLogger(NikoHomeControlCommunication1.class);
+
+    private final NhcSystemInfo1 systemInfo = new NhcSystemInfo1();
+    private final Map<String, NhcLocation1> locations = new ConcurrentHashMap<>();
+
+    @Nullable
+    private Socket nhcSocket;
+    @Nullable
+    private PrintWriter nhcOut;
+    @Nullable
+    private BufferedReader nhcIn;
+
+    private volatile boolean listenerStopped;
+    private volatile boolean nhcEventsRunning;
+
+    private ScheduledExecutorService scheduler;
+
+    // We keep only 2 gson adapters used to serialize and deserialize all messages sent and received
+    protected final Gson gsonOut = new Gson();
+    protected Gson gsonIn;
+
+    /**
+     * Constructor for Niko Home Control I communication object, manages communication with
+     * Niko Home Control IP-interface.
+     *
+     */
+    public NikoHomeControlCommunication1(NhcControllerEvent handler, ScheduledExecutorService scheduler) {
+        super(handler);
+        this.scheduler = scheduler;
+
+        // When we set up this object, we want to get the proper gson adapter set up once
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(NhcMessageBase1.class, new NikoHomeControlMessageDeserializer1());
+        this.gsonIn = gsonBuilder.create();
+    }
+
+    @Override
+    public synchronized void startCommunication() {
+        try {
+            for (int i = 1; nhcEventsRunning && (i <= 5); i++) {
+                // the events listener thread did not finish yet, so wait max 5000ms before restarting
+                Thread.sleep(1000);
+            }
+            if (nhcEventsRunning) {
+                logger.error("Niko Home Control: starting but previous connection still active after 5000ms");
```

Comment:
Please reduce log to `error` to `debug` Also prefer to put a text in the `IOException` below. In that case do you need to log it here? Or is it already logged by the method catching the `IOException`?

Reply:
If we get this, it would be a problem with the binding code, so I think error is the right level.  The IOException is caught in the same method with other IOException causes.
This code has not changed from the current binding, but due to refactoring it moved to NikoHomeControlCommunication1 rather than NikoHomeControlCommunication.

Reply:
https://www.openhab.org/docs/developer/development/guidelines.html#f-logging

The error level is not allowed for bindings except if you expect the entire openHAB to crash in a moment.

Reply:
This rule related to error was not strictly adhered to in the past, but we try to enforce it with new code. Unlucky you moved the code and this got marked as changed :slightly_smiling_face: See Davids comment about this logging guideline.
Something to think about, but not a blocking review: The fact it is a real problem with the binding is difficult to read from the code. Because it throws an `IOException`. Such an exception is normally an indication there is a problem with an external service and possible even something temporary. Just reading this code I would than expect the binding can recover from this error, by restarting itself or so?

Reply:
It actually happens when restarting the communication, and the previous communication has not closed properly yet (sockets seem to take some time to close). I wait for a maximum of 5 seconds before giving up and setting the bridge offline. I have never seen this after development of the binding 2 years ago, but did not want to keep something hanging.
I will change the log level.